sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -0,0 +1,334 @@
+        Lists.reverse(bundleDescriptor.getPrimitiveTransformList())) {
+      createConsumersForPrimitiveTransform(
+          primitiveTransform,
+          request::getInstructionId,
+          outputTargetToConsumer::get,
+          outputTargetToConsumer::put,
+          startFunctions::add,
+          finishFunctions::add);
+    }
+
+    // Already in reverse order so we don't need to do anything.
+    for (ThrowingRunnable startFunction : startFunctions) {
+      LOGGER.debug("Starting function {}", startFunction);
+      startFunction.run();
+    }
+
+    // Need to reverse this since we want to call finish in topological order.
+    for (ThrowingRunnable finishFunction : Lists.reverse(finishFunctions)) {
+      LOGGER.debug("Finishing function {}", finishFunction);
+      finishFunction.run();
+    }
+
+    return response;
+  }
+
+  /**
+   * Converts a {@link org.apache.beam.fn.v1.BeamFnApi.FunctionSpec} into a {@link DoFnRunner}.
+   */
+  private <InputT, OutputT> DoFnRunner<InputT, OutputT> createDoFnRunner(
+      BeamFnApi.FunctionSpec functionSpec,
+      Map<String, Collection<ThrowingConsumer<WindowedValue<OutputT>>>> outputMap) {
+    ByteString serializedFn;
+    try {
+      serializedFn = functionSpec.getData().unpack(BytesValue.class).getValue();
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(
+          String.format("Unable to unwrap DoFn %s", functionSpec), e);
+    }
+    DoFnInfo<?, ?> doFnInfo =
+        (DoFnInfo<?, ?>)
+            SerializableUtils.deserializeFromByteArray(serializedFn.toByteArray(), "DoFnInfo");
+
+    checkArgument(
+        Objects.equals(
+            new HashSet<>(Collections2.transform(outputMap.keySet(), Long::parseLong)),
+            doFnInfo.getOutputMap().keySet()),
+        "Unexpected mismatch between transform output map %s and DoFnInfo output map %s.",
+        outputMap.keySet(),
+        doFnInfo.getOutputMap());
+
+    ImmutableMultimap.Builder<TupleTag<?>,
+                              ThrowingConsumer<WindowedValue<OutputT>>> tagToOutput =
+                              ImmutableMultimap.builder();
+    for (Map.Entry<Long, TupleTag<?>> entry : doFnInfo.getOutputMap().entrySet()) {
+      tagToOutput.putAll(entry.getValue(), outputMap.get(Long.toString(entry.getKey())));
+    }
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    final Map<TupleTag<?>, Collection<ThrowingConsumer<WindowedValue<?>>>> tagBasedOutputMap =
+        (Map) tagToOutput.build().asMap();
+
+    OutputManager outputManager =
+        new OutputManager() {
+          Map<TupleTag<?>, Collection<ThrowingConsumer<WindowedValue<?>>>> tupleTagToOutput =
+              tagBasedOutputMap;
+
+          @Override
+          public <T> void output(TupleTag<T> tag, WindowedValue<T> output) {
+            try {
+              Collection<ThrowingConsumer<WindowedValue<?>>> consumers =
+                  tupleTagToOutput.get(tag);
+              if (consumers == null) {
```

Comment:
[no action needed] there's the Map lookup per output you were telling me about. :)